### PR TITLE
Add custom Serialize/Deserialize impl for EphemeralPublicKey, Pepper, Groth16ZkpAndStatement, G1Bytes, and G2Bytes, which output hex when used to serialize json

### PR DESCRIPTION
**Pull request by rex1fernando**
### Description

This PR adds Serialize/Deserialize implementations for the following types:

- EphemeralPublicKey
- Pepper
- Groth16ZkpAndStatement
- G1Bytes
- G2Bytes 

instead of using the derive macros. The purpose of this is so that serde_json serializes these two types as hex strings, while bcs behavior is still preserved.

### Test plan

This PR includes a new module `aptos_types::unit_tests::keyless_serialization_test`, which has test cases to test the new serialization/deserialization behavior for each of the types above.

### CODE REVIEW

**Review by alinush — APPROVED** [review]
Be sure to run `./scripts/rust_lint.sh` since, judging from the `authenticator.rs` changes, I think you might've linted with an older `cargo` or `rustc`...

**Review by heliuchuan — APPROVED** [review]

File: types/src/transaction/authenticator.rs [hunk 1]
```diff
@@ -1121,6 +1130,58 @@ impl TryFrom<&[u8]> for EphemeralPublicKey {
...
+
+            let value = Value::deserialize(deserializer)?;
+            Ok(match value {
+                Value::Ed25519 { public_key } => EphemeralPublicKey::Ed25519 { public_key },
+            })
+        }
+    }
+}
+
+impl Serialize for EphemeralPublicKey {
```
[the remark this answers is not in the document]
Comment (rex1fernando) [reply types/src/transaction/authenticator.rs:1159]: @heliuchuan added a new module for tests, with one testcase per modified type. Let me know what you think

File: types/src/keyless/bn254_circom.rs [hunk 2]
```diff
@@ -62,6 +62,18 @@ impl G1Bytes {
         Self::new_from_vec(bytes)
     }
 
+    // TODO(keyless) Shouldn't this return a result rather than unwrap()ing?
+    pub fn from_hex(hex: &str) -> Self {
```
Comment (alinush) [inline types/src/keyless/bn254_circom.rs:66]: Remove these functions and use them directly in the Serialize/Deserialize impl block

File: types/src/keyless/mod.rs [hunk 3]
```diff
@@ -142,10 +145,12 @@ impl KeylessSignature {
 ///
 /// This value should **NOT* be changed since on-chain addresses are based on it (e.g.,
 /// hashing with a larger pepper would lead to a different address).
-#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
 pub struct Pepper(pub(crate) [u8; poseidon_bn254::BYTES_PACKED_PER_SCALAR]);
 
 impl Pepper {
+    // TODO(keyless) Account address uses Self::LENGTH instead of Self::NUM_BYTES. Consider
```
Comment (alinush) [inline types/src/keyless/mod.rs:152]: LENGTH is not specific enough: length in bytes ? words? bits?

File: types/src/keyless/bn254_circom.rs [hunk 4]
```diff
@@ -114,6 +162,18 @@ impl G2Bytes {
         Self::new_from_vec(bytes)
     }
 
+    // TODO(keyless) Shouldn't this return a result rather than unwrap()ing?
```
Comment (alinush) [inline types/src/keyless/bn254_circom.rs:165]: Remove these too

File: types/src/keyless/mod.rs [hunk 5]
```diff
@@ -156,13 +161,18 @@ impl Pepper {
...
 
+    // TODO(keyless) Shouldn't this return a result rather than unwrap()ing?
     pub fn from_hex(hex: &str) -> Self {
         let bytes = hex::decode(hex).unwrap();
         let mut extended_bytes = [0u8; Self::NUM_BYTES];
         extended_bytes.copy_from_slice(&bytes);
         Self(extended_bytes)
     }
 
+    pub fn to_hex(&self) -> String {
```
Comment (alinush) [inline types/src/keyless/mod.rs:172]: remove this

File: types/src/transaction/authenticator.rs [hunk 6]
```diff
@@ -1110,6 +1110,15 @@ impl EphemeralPublicKey {
     pub fn to_bytes(&self) -> Vec<u8> {
         bcs::to_bytes(self).expect("Only unhandleable errors happen here.")
     }
+
+    pub fn from_hex(hex: &str) -> Result<Self, CryptoMaterialError> {
```
Comment (alinush) [inline types/src/transaction/authenticator.rs:1114]: remove these

File: types/src/unit_tests/keyless_serialization_test.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    assert_eq!(
+        serde_json::to_string(&serde_json::from_str::<EphemeralPublicKey>(epk_str).unwrap())
+            .unwrap()
+            .as_str(),
+        epk_str
+    );
+    assert_eq!(
+        bcs::to_bytes(&bcs::from_bytes::<EphemeralPublicKey>(&epk_bytes).unwrap()).unwrap(),
+        epk_bytes
+    );
```
Comment (alinush) [inline types/src/unit_tests/keyless_serialization_test.rs:46]: Can you add an assert on the BCS serialized length? It should be 32 bytes + 2 bytes I believe.

File: types/src/unit_tests/keyless_serialization_test.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    assert_eq!(
+        serde_json::to_string(&serde_json::from_str::<Pepper>(pepper_str).unwrap())
+            .unwrap()
+            .as_str(),
+        pepper_str
+    );
+    assert_eq!(
+        bcs::to_bytes(&bcs::from_bytes::<Pepper>(&pepper_bytes).unwrap()).unwrap(),
+        pepper_bytes
+    );
```
Comment (alinush) [inline types/src/unit_tests/keyless_serialization_test.rs:80]: Like before, assert BCS serialized length is correct.

File: types/src/unit_tests/keyless_serialization_test.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        .as_str(),
+        groth16_zkp_and_statement_str
+    );
+    assert_eq!(
+        bcs::to_bytes(
+            &bcs::from_bytes::<Groth16ZkpAndStatement>(&groth16_zkp_and_statement_bytes).unwrap()
+        )
+        .unwrap(),
+        groth16_zkp_and_statement_bytes
+    );
```
Comment (alinush) [inline types/src/unit_tests/keyless_serialization_test.rs:139]: Like before, assert BCS serialized length is correct.

**Pull request merged**